### PR TITLE
Switch OrganizationGithubAppInstallation away from ja_resource and canary

### LINKED ISSUE #893: Switch OrganizationGithubAppInstallationController away from ja_resource and canary
Progress on #864 

# Problem

We want to move away from the `ja_resource` and `canary` combo used in most of our controller modules. The central issue for this project is #864

There is a generic process of how to do it in that issue, as well as tracking for the milestone progress. 

The procedure is also pasted below. If this specific controller deviates from this standard procedure, feel free to ask for help here, or by opening a PR.

The PR should indicate that it's closing this issue and is progressing #864 further along.

# Procedure

- pick a controller to switch
- view actions this controller should support in `lib/code_corps_web/router.ex`
- add `action_fallback CodeCorpsWeb.FallbackController`. This makes it so in case a with clause within the controller action fails, a sort of "catch all" function will be called from `FallbackController`
- add `plug CodeCorpsWeb.Plug.DataToAttributes`. This converts the `data` hash within the conn params from the json api structure into a flat map convenient for usage with changesets
- for each action
   - remove that action from any `load_resource`, `load_and_authorize_resource`, etc. plugs in that controller
   - remove any `def handle_#{action}` functions
   - write a `def #{action}(conn, params) do`. For the standard `create`, `update` `delete`, `show` and index actions. See examples below on how these functions should be written, usually
   - rewrite the policy function for that action to support params instead of a changeset if necessary. See section below for pointers.

That should be the whole procedure. We already have controller tests in place, so if the process has been done correctly, the tests should pass.

Some additional steps might be necessary, however, if there are some specifics to an action. For those, we will be here to provide feedback.

## How to rewrite controller actions

```Elixir
@spec index(Conn.t, map) :: Conn.t
def index(%Conn{} = conn, %{} = params) do
  with resources <- ResourceModule |> Query.id_filter(params) |> Repo.all do
    conn |> render("index.json-api", data: resources)
  end
end

@spec show(Conn.t, map) :: Conn.t
def show(%Conn{} = conn, %{"id" => id}) do
  with %ResourceModule{} = resource <- ResourceModule |> Repo.get(id) do
    conn |> render("show.json-api", data: resource)
  end
end

@spec create(Plug.Conn.t, map) :: Conn.t
def create(%Conn{} = conn, %{} = params) do
  with %User{} = current_user <- conn |> Guardian.Plug.current_resource,
       {:ok, :authorized} <- current_user |> Policy.authorize(:create, params),
       {:ok, %Comment{} = resource} <- %ResourceModule{} |> ResourceModule.create_changeset(params) |> Repo.insert do
    conn |> put_status(:created) |> render("show.json-api", data: resource)
  end
end

@spec update(Conn.t, map) :: Conn.t
def update(%Conn{} = conn, %{"id" => id} = params) do
  with %ResourceModule{} = resource <- ResourceModule |> Repo.get(id),
       %User{} = current_user <- conn |> Guardian.Plug.current_resource,
       {:ok, :authorized} <- current_user |> Policy.authorize(:update, resource),
       {:ok, %ResourceModule{} = resource} <- resource |> ResourceModule.changeset(params) |> Repo.update do
    conn |> render("show.json-api", data: resource)
  end
end
```

So to recap, the actions work like

### `index` action

- apply filters from query params to load data
- render data

### `show` action

- load resource
- render resource

### `create` action

- authorize action for current user and parameters (requires rewriting that policy to use a params map instead of a changeset, expect `{:ok, :authorized}`
- perform insert action
- render resulting resource

### `update` and `delete` actions

- load resource
- authorize action for current user on resource, expect `{:ok, :authorized}`
- perform update/delete
- render resource, or no content for delete

### Examples in code

-`CodeCorpsWeb.CommentController` has examples for standard `index`, `show`, `create` and `update` actions

## How to rewrite policies

This consists of several relatively simple steps.

- For each action being switch away, we need to move the `def can` function clause out of the `defimpl` block in `policy.ex` into the root namespace of the `CodeCorps.Policy` module. 
- We also need to make the clause private instead of public.

- If any of the clauses expect a changeset, they should now expect a simple map of params:

```Elixir
defp can?(%User{} = user, :some_action, %Changeset{data: %Resource{}} = changeset), do: Policy.Resource.some_action?(user, changeset)
```

becomes

```Elixir
defp can?(%User{} = user, :some_action, %{} = params), do: Policy.Resource.some_action?(user, params)
```

- in this case, the policy module itself also need to be rewritten to support a params map. Since in most cases, these policies simply look at the `changeset.changes` map, the rewrite ought to be simpler. If there is trouble, someone will be available to provide pointers.

### Examples in code

- the rewrite has been done for `Policy.Comment.create?`

### PR DESCRIPTION
**Pull request by treble37**
Switch OrganizationGithubAppInstallationController away from ja_resource and canary

Closes #893

# What's in this PR?

Switch OrganizationGithubAppInstallationController away from ja_resource and canary

Make sure any changes to code include changes to documentation.

## References
Fixes # 893

Progress on: # 864

### CODE REVIEW

File: lib/code_corps_web/controllers/organization_github_app_installation_controller.ex [hunk 1]
```diff
@@ -22,9 +16,45 @@ defmodule CodeCorpsWeb.OrganizationGithubAppInstallationController do
...
+      {:ok, %OrganizationGithubAppInstallation{} = organization_installation} <- organization_installation |> OrganizationGithubAppInstallation.changeset(params) |> Repo.update do
+        conn |> render("show.json-api", data: organization_installation)
+    end
+  end
+
+  @spec delete(Plug.Conn.t, map) :: Conn.t
+  def delete(%Conn{} = conn, %{} = params) do
+    with %User{} = current_user <- conn |> Guardian.Plug.current_resource,
+         {:ok, :authorized} <- current_user |> Policy.authorize(:delete, %OrganizationGithubAppInstallation{}, params),
+         {:ok, %OrganizationGithubAppInstallation{} = organization_installation} <- %OrganizationGithubAppInstallation{} |> OrganizationGithubAppInstallation.delete_changeset(params) |> Repo.insert do
```
Comment (joshsmith) [inline lib/code_corps_web/controllers/organization_github_app_installation_controller.ex]: Looks like the build is erroring because `OrganizationGithubAppInstallation.delete_changeset/2` doesn't exist. I'm not sure this needs a changeset, just a delete operation.

Comment (treble37) [reply lib/code_corps_web/controllers/organization_github_app_installation_controller.ex]: Thanks @JoshSmith, the tests were running locally (3 failures) so I didn't notice the build and I've updated...

**Comment by treble37** [comment]
I get 2 failing tests locally now with message like:

```
1) test delete deletes resource (CodeCorpsWeb.OrganizationGithubAppInstallationControllerTest)
     test/lib/code_corps_web/controllers/organization_github_app_installation_controller_test.exs:75
     ** (RuntimeError) expected response with status 204, got: 403, with body:
     {"errors":[{"title":"403 Forbidden","status":403,"id":"FORBIDDEN","detail":"You are not authorized to perform this action."}]}
     stacktrace:
       (phoenix) lib/phoenix/test/conn_test.ex:362: Phoenix.ConnTest.response/2
       test/lib/code_corps_web/controllers/organization_github_app_installation_controller_test.exs:79: (test)
```

I wonder how I setup the authorization correctly?

File: lib/code_corps_web/controllers/organization_github_app_installation_controller.ex [hunk 2]
```diff
@@ -22,9 +16,45 @@ defmodule CodeCorpsWeb.OrganizationGithubAppInstallationController do
...
+        conn |> render("show.json-api", data: organization_installation)
+    end
+  end
+
+  @spec delete(Plug.Conn.t, map) :: Conn.t
+  def delete(%Conn{} = conn, %{} = params) do
+    with %User{} = current_user <- conn |> Guardian.Plug.current_resource,
+         {:ok, :authorized} <- current_user |> Policy.authorize(:delete, %OrganizationGithubAppInstallation{}, params),
+         {:ok, %OrganizationGithubAppInstallation{} = organization_installation} <- %OrganizationGithubAppInstallation{} |> OrganizationGithubAppInstallation.changeset(params) |> Repo.insert do
+      conn |> put_status(:deleted) |> render("show.json-api", data: organization_installation)
```
Comment (begedin) [inline lib/code_corps_web/controllers/organization_github_app_installation_controller.ex]: The order of operations here should be

- load organization github app installation by id
- `{:ok, :authorized} <- current_user |> Policy.authorize(:delete, organization_github_app_installation)`,
- `{:ok, %OrganizationGithubAppInstallation{}} <- organization_github_app_installation |> Repo.delete`
- render a 204 (basically, `conn |> send_resp(:no_content, "")`

If you implement it that way, you do should 
- not have to make the changeset function public
- implement the delete policy to accept a record, not params

File: lib/code_corps_web/controllers/organization_github_app_installation_controller.ex [hunk 3]
```diff
@@ -22,9 +16,45 @@ defmodule CodeCorpsWeb.OrganizationGithubAppInstallationController do
...
+
+  @spec update(Conn.t, map) :: Conn.t
+  def update(%Conn{} = conn, %{"id" => id} = params) do
+    with %OrganizationGithubAppInstallation{} = organization_installation <- OrganizationGithubAppInstallation |> Repo.get(id),
+      %User{} = current_user <- conn |> Guardian.Plug.current_resource,
+      {:ok, :authorized} <- current_user |> Policy.authorize(:update, organization_installation),
+      {:ok, %OrganizationGithubAppInstallation{} = organization_installation} <- organization_installation |> OrganizationGithubAppInstallation.changeset(params) |> Repo.update do
+        conn |> render("show.json-api", data: organization_installation)
+    end
+  end
```
Comment (begedin) [inline lib/code_corps_web/controllers/organization_github_app_installation_controller.ex]: We don't actually support an `update` in this controller, so this action should be completely removed.

A good way to determine which are supported is to look into the `router.ex`:

```Elixir
# in the non-authenticated scope, we have

resources "/organization-github-app-installations", OrganizationGithubAppInstallationController, only: [:index, :show]

# in the authenticated scope, we have

resources "/organization-github-app-installations", OrganizationGithubAppInstallationController, only: [:create, :delete]
```

That makes the supported actions only `[:index, :show, :create, :delete]`

If you think about it, it makes sense. An `organization_github_app_installation` is a many-to-many link between an `organization` and a `github-app-installation`, so there is nothing to change about an existing record. It either exists, or it doesn't.

File: lib/code_corps/policy/policy.ex [hunk 4]
```diff
@@ -31,6 +31,8 @@ defmodule CodeCorps.Policy do
   defp can?(%User{} = user, :create, %Organization{}, %{}), do: Policy.Organization.create?(user)
   defp can?(%User{} = user, :update, %Organization{} = organization, %{}), do: Policy.Organization.update?(user, organization)  
   defp can?(%User{} = current_user, :update, %User{} = user, %{}), do: Policy.User.update?(user, current_user)
+  defp can?(%User{} = user, :delete, %OrganizationGithubAppInstallation{}, %{} = params), do: Policy.OrganizationGithubAppInstallation.delete?(user, params)
```
Comment (begedin) [inline lib/code_corps/policy/policy.ex]: If you fix the `delete` action below, this will become

```Elixir
defp can?(%User{} = user, :delete, %OrganizationGithubAppInstallation{} = organization_github_app_installation, %{}), do: Policy.OrganizationGithubAppInstallation.delete?(user, organization_github_app_installation)
```

File: lib/code_corps/policy/organization_github_app_installation.ex [hunk 5]
```diff
@@ -6,9 +6,19 @@ defmodule CodeCorps.Policy.OrganizationGithubAppInstallation do
...
-  def create?(%User{} = user, %Ecto.Changeset{} = changeset),
-    do: changeset |> get_organization |> owned_by?(user)
+  def create?(%User{} = user, params) do
+    params |> atomize_params |> get_organization |> owned_by?(user)
+  end
 
-  def delete?(%User{} = user, %OrganizationGithubAppInstallation{} = github_app_installation),
-    do: github_app_installation |> get_organization |> owned_by?(user)
+  def delete?(%User{} = user, params),
+    do: params |> atomize_params |> get_organization |> owned_by?(user)
```
Comment (begedin) [inline lib/code_corps/policy/organization_github_app_installation.ex]: You can completely revert this change if you fix the controller action.

**Review by begedin — CHANGES_REQUESTED** [review]
Some changes to be made, but otherwise, good first draft. 👍

**Comment by joshsmith** [comment]
Hey @treble37 just a heads up that `lib/code_corps/policy/policy.ex` has changed so you'll need to rebase on upstream `develop` and bring in the new changes from upstream.

Let me know if you need any help at all.

**Comment by joshsmith** [comment]
Hey @treble37 wanted to check in on this again and see if you need anything from me.

**Comment by treble37** [comment]
Thanks @JoshSmith, I just pulled down the latest dev branch changes, will rebase and get back to you...it's just been a busy week

File: lib/code_corps_web/controllers/organization_github_app_installation_controller.ex (hunk 2 above)
Comment (treble37) [reply lib/code_corps_web/controllers/organization_github_app_installation_controller.ex]: Updated @begedin, thanks

File: lib/code_corps/policy/policy.ex (hunk 4 above)
Comment (treble37) [reply lib/code_corps/policy/policy.ex]: updated

File: lib/code_corps/policy/organization_github_app_installation.ex (hunk 5 above)
Comment (treble37) [reply lib/code_corps/policy/organization_github_app_installation.ex]: updated

**Comment by treble37** [comment]
@JoshSmith - `mix test` passes locally but not on circle CI:

```
....................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................

Finished in 14.5 seconds
1164 tests, 0 failures
```

**Comment by begedin** [comment]
@treble37 I restarted the build and the tests passed. It must have been some parallelism issue with how bypass uses ports or something.

We need you to squash your commits into one and rebase before we can merge.

**Comment by treble37** [comment]
squashed @begedin , thanks!

**Review by begedin — APPROVED** [review]

File: lib/code_corps_web/controllers/organization_github_app_installation_controller.ex [hunk 6]
```diff
@@ -1,18 +1,12 @@
...
-  plug :load_resource, model: OrganizationGithubAppInstallation, only: [:show], preload: @preloads
-  plug :load_and_authorize_changeset, model: OrganizationGithubAppInstallation, only: [:create], preload: @preloads
-  plug :load_and_authorize_resource, model: OrganizationGithubAppInstallation, only: [:delete]
-
-  plug JaResource
+  action_fallback CodeCorpsWeb.FallbackController
+  plug CodeCorpsWeb.Plug.DataToAttributes
 
   @spec model :: module
   def model, do: CodeCorps.OrganizationGithubAppInstallation
```
Comment (begedin) [inline lib/code_corps_web/controllers/organization_github_app_installation_controller.ex]: Sorry I missed this on the first pass, but the method here is no longer necessary now, so you can remove it.

So I don't annoy you further, don't forget to squash again, as well as run a `git pull develop` followed by `git rebase origin/develop` at the end of it. Then I can merge.

**Review by begedin — CHANGES_REQUESTED** [review]

File: lib/code_corps_web/controllers/organization_github_app_installation_controller.ex (hunk 6 above)
Comment (treble37) [reply lib/code_corps_web/controllers/organization_github_app_installation_controller.ex]: ok thanks for your patience @begedin with my first contribution, I rebased on the latest origin/develop and removed `def model` method....thanks!

Comment (treble37) [reply lib/code_corps_web/controllers/organization_github_app_installation_controller.ex]: ok thanks for your patience @begedin with my first contribution, I rebased on the latest origin/develop and removed def model method....thanks!

Comment (begedin) [reply lib/code_corps_web/controllers/organization_github_app_installation_controller.ex]: I'm the one who should be thankful @treble37 

And now I have to apologize again and ask you to do yet another change. The `filter` method also isn't required anymore. Mind removing that one to? 

Again, squash and rebase after.

I'm really sorry about this, but the "changes" view on github really sucks when it comes to seeing which code needs to be removed.

Comment (treble37) [reply lib/code_corps_web/controllers/organization_github_app_installation_controller.ex]: No worries @begedin, filter removed!

**Review by begedin — APPROVED** [review]

**Comment by begedin** [comment]
@treble37 Excellent work! 👍

**Comment by joshsmith** [comment]
🙌
